scripts/src/java/org/oppia/android/scripts/coverage/RunCoverage.kt

```diff
@@ -56,8 +55,8 @@ const val BOLD = "\u001B[1m"
  *
  * Example with output path to save the collected coverage proto:
  *    bazel run //scripts:run_coverage -- $(pwd)
- *    utility/src/main/java/org/oppia/android/util/parser/math/MathModel.kt
- *    --protoOutputPath=/tmp/coverage_report.proto64
+ *    utility/src/main/java/org/oppia/android/util/parser/math/MathModel.kt --format=PROTO
+ *    --protoOutputPath=/tmp/coverage_report.pb
```

Comment:
Part of the reason I suggested using a PROTO output format is that you can avoid needing to pass in a path (the only time I can think of that we *must* pass in a path is when writing custom Bazel rules, so that doesn't apply here).

You could, instead, make some sort of predictable output path for all of the different report types. For example:

coverage_reports/utility/src/main/.../MathModel.kt/coverage_report.[pb/md/html]

This makes it easier when using the script as a user, but it also creates a predictable location for other scripts to read/retrieve these files.

Reply:
Sorry if I’m off track, but I was considering introducing an output path because we’ll need it for CI. My initial plan was to run coverage in CI as shards, and with each shard's results saved as `coverage_report_shard***.pb` (most probably planning to have the encoded shard names as suffix - Although this approach hasn’t been implemented yet and might change if it doesn’t work as expected, would be implementing this today), the idea is that once all shards have finished running,

```mermaid
graph LR
    A[Coverage Run Shard 1] --> B1[Coverage Report Shard 1.pb]
    A2[Coverage Run Shard 2] --> B2[Coverage Report Shard 2.pb]
    A3[Coverage Run Shard 3] --> B3[Coverage Report Shard 3.pb]

    B1 --> C[Report Combiner Script]
    B2 --> C
    B3 --> C

    C --> D[Coverage Reporter]

    style A fill:#ccf,stroke:#333,stroke-width:2px
    style A2 fill:#ccf,stroke:#333,stroke-width:2px
    style A3 fill:#ccf,stroke:#333,stroke-width:2px
    style B1 fill:#cfc,stroke:#333,stroke-width:2px
    style B2 fill:#cfc,stroke:#333,stroke-width:2px
    style B3 fill:#cfc,stroke:#333,stroke-width:2px
    style C fill:#ff9,stroke:#333,stroke-width:2px
    style D fill:#f9f,stroke:#333,stroke-width:4px
```

we would collect all the .pb files and use a separate job with a script to merge them into a single coverage report container. This combined report would then be sent to the coverage reporter to generate a Markdown report.

I’m wondering if we can skip the output path, how can we determine the file path for CI. For local development, we could specify a predictable path with its own individual file path (similar to how HTML reports are handled), but for CI, what would be the preferred method for saving these files?

Also, for md, we would be having just one common report so chose a common location for it and (in future pr for uploading the comments (called as script from ci)) will be passing a location for it too to save it in ci runs and later use that to upload it as comment. It's only for HTML (local dev purposes) we have distinct paths.

Reply:
The common MD comes from generating a report from the combined list proto, right? The MD suggestion here is just for when a developer selects to generate Markdown when running the script locally.

I was thinking that you could collect all the pb files by just iterating in the generated coverage_reports directory for all *.pb files--this is straightforward to do for both shell scripts (e.g. using ``find``) and in Kotlin.

Reply:
Got it, thanks for the suggestion @BenHenning, implemented saving the protos for each file in their respective folders based on their filepaths.

**with --format=PROTO**
```kotlin
app/src/main/java/org/oppia/android/app/activity/ActivityComponent.kt

/* would save it to */
./coverage_reports/app/src/main/java/org/oppia/android/app/activity/ActivityComponent/coverage_report.pb

```

And also updated the test to test this - testRunCoverage_withProtoReportFormat_savesCoverageReportProto

#

>The common MD comes from generating a report from the combined list proto, right?

"**Yes**", the common MD is for the final combined proto.

>The MD suggestion here is just for when a developer selects to generate Markdown when running the script locally.

But still the need to save md reports based on file name paths is quite unclear to me. Do we need to save each and every md report? Would that be useful to have something like:

```
| File.kt | 13% | 10 /  54 | :x: | 70% |
```

in a separate file, as the above will be the md report for a single file run, and thought md wouldn't add much value to local dev (even if we save it in a complete table form) and skipped saving them as individual reports and only made effort to save them in the final stage (for ci and also for local dev debugging) (probably I will add a print statement to the log the final md report path - which is right now missing (html reports do log them)).

Reply:
I actually don't think there's a strong imperative to match the pathing for the MD file other than consistency since it's only for local development. The additional path output line was a nice idea to add; I think the current state of the script is in a good state for local development so I don't have any strong feelings to make additional changes here.